### PR TITLE
Fix write operations potentially using  other-threaded tracked instances

### LINKED ISSUE #1427: Crash on attempting to perform beatmap operations
Now that EF contexts are per-thread, `Hide`/`Delete`/`Restore`/`Undelete` operations need to refresh the input to the local context before making changes. Currently they may receive a beatmap from a different thread's context, which can result in the following exception:

`The instance of entity type 'BeatmapInfo' cannot be tracked because another instance with the same key value for {'ID'} is already being tracked. When attaching existing entities, ensure that only one entity instance with a given key value is attached. Consider using 'DbContextOptionsBuilder.EnableSensitiveDataLogging' to see the conflicting key values.`

Reproduction:
- enter song select
- exit song select
- enter song select
- hide a beatmap difficulty

### PR DESCRIPTION
**Pull request by peppy**
We need to ensure we have a tracked instance from the local thread's DbContext before performing write operations, else exceptions may be thrown.

Resolves #1427.

### CODE REVIEW

File: osu.Game/Database/DatabaseBackedStore.cs [hunk 1]
```diff
@@ -19,6 +22,22 @@ public abstract class DatabaseBackedStore
         private readonly ThreadLocal<OsuDbContext> queryContext;
 
         /// <summary>
+        /// Refresh an instance potentially from a different thread with a local context-tracked instance.
+        /// </summary>
+        /// <param name="obj"></param>
+        /// <param name="lookupSource"></param>
```
Comment (UselessToucan) [inline osu.Game/Database/DatabaseBackedStore.cs]: Would be nice to have some explanation here

Comment (peppy) [reply osu.Game/Database/DatabaseBackedStore.cs]: yep, will populate these.

File: osu.Game/Database/DatabaseBackedStore.cs [hunk 2]
```diff
@@ -19,6 +22,22 @@ public abstract class DatabaseBackedStore
...
+        /// <typeparam name="T"></typeparam>
+        protected virtual void Refresh<T>(ref T obj, IEnumerable<T> lookupSource = null) where T : class, IHasPrimaryKey
+        {
+            var context = GetContext();
+
+            if (context.Entry(obj).State != EntityState.Detached) return;
+
+            var id = obj.ID;
+            obj = lookupSource?.FirstOrDefault(t => t.ID == id) ?? context.Find<T>(id);
```
Comment (UselessToucan) [inline osu.Game/Database/DatabaseBackedStore.cs]: Since ID is unique, it is safe to use SingleOrDefault here

**Review by smoogipoo — DISMISSED** [review]
After discussion, this looks sane to me.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**